### DISCUSSION TITLE
Roland R-8m Drum Modul Panel, need a little help from you!

### DISCUSSION THREAD

**Discussion by hps2109** [RomanKubiak/ctrlr#673] [General]
Hello to everyone,

btw. the Roland R-8m is is a parameter monster, so it makes sense to be able to set the parameters clearly with a panel, instead of using 6 buttons for every key note… uuuaaahhhh :-)
Roudabout 800 parameter or more for one patch, including 4* Performance and possibly additionally
8 * Feel Patches. I really hope, that CTRL can manage this :-)

So far the panel works as I wanted, but before I continue I have a question to the LUA experts of you.

All parameters are directly sent from the modulators via sysex, including the Roland checksum, also those that send negative values. All works perfekt! :-)

All mods are set to the correct values per sysex patch receive.

Except!!! all the modulators with negative values…(disabled at the moment) :-(

Sure, I also searched the old ctrl forum and various panels too. But, unfortunately, always affects to be „send“ the negative values, ​​but not what to do via Lua, when receiving this values per SysEx, splitted in lsb and msb bytes.

How can I set these values to the correct modulator values? I just can't figure it out :-(
I'm sure, someone of you has an idea for me. That would be really, really helpful for me.

The modulator "Pitch" has a value range from -480 to 480, where the "0" value is at position "480".
Example for a received SysEx string: Pitch for Key C2 (**24h**) Instrument:
F0 41 00 36 12 01 24 02 **20** 39 F7   LSB  -480  (Mod Value must be = 0 to set)
F0 41 00 36 12 01 24 03 **7C** 5C F7  MSB -480  (Mod Value must be = 0 to set)

F0 41 00 36 12 01 24 02 **00** 59 F7   LSB   0      (Mod Value must be = 480 to set)
F0 41 00 36 12 01 24 03 **00** 58 F7   MSB  0      (Mod Value must be = 480 to set)

F0 41 00 36 12 01 24 02 **60** 79 F7   LSB  480   (Mod Value must be = 960 to set)
F0 41 00 36 12 01 24 03 **03** 55 F7  MSB  480  (Mod Value must be = 960 to set)

Kind regards and many, many thanks for your help
Peter

![Key C2-B2](https://github.com/user-attachments/assets/5354d150-c3ca-47e4-86ff-5901795413ae)

![Performance](https://github.com/user-attachments/assets/9feb6dc2-9163-485d-a10b-c186b180fe0c)

**Comment by hps2109** [comment]
btw. does anyone have a horizontal slider for me? the "KnobGallery" at www.g200kg.com link for png's is dead, isn't it?

**Reply by damiensellier** [reply]
you can use wayback machine to recover old web pages : 
https://web.archive.org/web/20220825043508/http://www.g200kg.com/en/webknobman/gallery.php

**Reply by hps2109** [reply]
Hi Damien, thanks a lot for this tip!

**Comment by dervur** [comment]
Is pitch a CC? 
I never worked with sysex so I might be wrong, but you can't have negative value from my understanding of MIDI... Again I might be wrong.
I was doing a panel for the Volca Sample and the speed value goes from -63 to 63, but to do this behaviour I had to set a slider and write all the values from -63 to 63, but the CC message still goes from 0 to 127 at the output, before it wouldn't send...

I might be talking bullshit here cause I don't know sysex though so take what I say with a grain of salt

**Reply by damiensellier** [reply]
pitch relies on 2 bytes. One for coarse, one for fine. MSB LSB alike. It gives more precision like this. 

_In the pitchbend message, we combine the two bytes (the LSB and the MSB) to make a single 14-bit value that has a range from 0 to 16,383. We do that by bit-shifting the MSB 7 bits to the left and combining that with the LSB using a bitwise OR operation (or by addition). So, for example, if we receive a MIDI message “224 120 95” that means “pitchbend on channel 1 with a coarse setting of 95 and a fine resolution of 120 (i.e., 120/128 of the way from 95 to 96)”. If we bit-shift 95 (binary 1011111) to the left by 7 bits we get 12,160 (binary 10111110000000), and if we then combine that with the LSB value 120 (binary 1111000) by a bitwise OR or by addition, we get 12,280 (binary 10111111111000).

The MIDI protocol specifies that a pitchbend value of 8192 (MSB of 64 and LSB of 0) means no bend. Thus, on the scale from 0 to 16,383, a value of 0 means maximum downward bend, 8,192 means no bend, and 16,383 means maximum upward bend. Almost all pitchbend wheels on MIDI controllers use a spring mechanism that has the dual function of a) providing tactile resistance feedback as one moves the wheel away from its centered position and b) snapping the wheel quickly back to its centered position when it’s not being manipulated._

https://sites.uci.edu/camp2014/2014/04/30/managing-midi-pitchbend-messages/

**Reply by hps2109** [reply]
Hi Damien,

thank you very much for the tip about representing the numbers in binary!
That really makes sense, especially we can see it directly visual.
I have no idea, why I didn't think of that...

Btw. I saw that the old CTRLR Forum is no longer available. 
I had saved so many interesting links as information :-(
Is it possible to print a variable as binary in the console?

Okay, I have combined all the nibbles and bits and see the result as hex.
Roland managed the higher pitch bits across different nibbles... uahhh.

Byte Offset
0-1		00 00 	Media lsb bit0-3, Media MSB bit0, VeloCurve bit1-3 (0-7)
2-3		00 00 	Inst lsb bit0-3, Inst MSB bit0, NoteOff RX bit3 (0-1)
4	        00            **Pitch lsb             > bit0-3**
5              00 	        **Pitch                   > bit 4-7**
6		00            Decay1 lsb  bit0-3
7              00 	        Decay1 msb bit0-2
			        **Pitch bit3            > bit 8**
8		00           Decay2 lsb bit0-3
9              00 	       Decay2 MSB bit0-2
			       **Pitch msb bit3  - >  bit 9**
10		00 	       Level     bit0-3	(0-15)
11		08 	       Nuance bit0-3	(0-15)
12		03 	       Out Assign	bit0-3 (3 = STEREO)
13		09 	       Assign Type 	bit0-3 (9 = POLY)

Okay, I combined the nibbles & bits within LUA.
Sure, there is room for improvement, but it works for now :-)

    -- Pitch
       local getBit03 = parameterData1:getByte(112+step)      -- get nibble bit 0-3 0000xxxx (lsb)
       local getBit47 = parameterData1:getByte(113+step)      -- get nibble bit 4-7 0000yyyy
       local getBit8 = parameterData1:getByte(115+step)        -- get nibble incl. bit 8    0000z000
       local getBit9 = parameterData1:getByte(117+step)        -- get nibble incl. bit 9    0000w000 (msb)
 
       pitchBit8 = bit.rshift(tonumber(getBit8),3)                    -- shift bit 8 to  0000000z
       pitchBit9 = bit.rshift(tonumber(getBit9),3)                    -- shift bit 9 to  0000000w

       console(string.format("Bit9: %.2x, Bit8: %.2x, Bit7-4: %.2x, Bit3-0: %.2x",pitchBit9,pitchBit8,getBit47,getBit03))
       -- show Bits & nibbles in hex 

       local pitchValue = assemble4bitNibble(getBit47, getBit03)     -- Function assemble4bitNibble

       if pitchBit9 == 0 then
          pitchValue = pitchValue + 480 + pitchBit8*256
--          console("Pitch Mod Value+ :"..pitchValue)
          panel:getModulatorByName(selectPitch):setModulatorValue(pitchValue, true, false, true)
       elseif pitchBit9 == 1 then
          pitchValue = pitchValue + 480 + pitchBit8*256 - 512
--          console("Pitch Mod Value- :"..pitchValue)
          panel:getModulatorByName(selectPitch):setModulatorValue(pitchValue, true, false, true)
       end

      function assemble4bitNibble(a, b)      -- convert two 4 bit nibbles to a single integer
          local m = bit.lshift(a, 4)
          return m + b
     end


Kind regards
Peter

**Comment by damiensellier** [comment]
Hi Peter,
To display your binary value in the console, with LUA you can use the usual string.tohex() and string.fromhex() functions: 

```
function string.fromhex(str)
    return (str:gsub('..', function (cc)
        return string.char(tonumber(cc, 16))
    end))
end
```

```
function string.tohex(str)
    return (str:gsub('.', function (c)
        return string.format('%02X', string.byte(c))
    end))
end
```
Hope it helps
Damien

**Reply by hps2109** [reply]
Hi Damien, many thanks for the functions. I'll try it out.

**Comment by dnaldoog** [comment]
This is one way to do it using a _uiSlider_ - You can also use a _uiFixedSlider_.

Set your _uiSlider_ to -480 +480. 

I will name it **pitch**

Set this field like this and hit enter to register the expression:

![image](https://github.com/user-attachments/assets/5ed3209e-cc4f-4f7b-b945-c670fb19342b)

Set the MIDI to sysex with a formula like  `F0 41 00 36 12 01 24 MS LS z5 F7`

In this field you need a call back function for received MIDI:

![image](https://github.com/user-attachments/assets/b9cbbef0-8ebf-4773-bca5-40ed256c2dbe)

The lua function in this callback will be something like:
<pre>
midiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    if midi:getSize() == 11 then
        local MSB = midi:getData():getByte(7) * 128
        local LSB = midi:getData():getByte(8)
        local sum = MSB + LSB
        sum = sum - 480
        panel:getModulatorByName("pitch"):setModulatorValue(sum, false, false, false)
    end
end
</pre>
Roland usually uses zero for the lowest negative `00 00` so you need to subtract 480 in this case to get to -480 **on the slider**  and to get to +480 subtract 480 fron 960 which in two bytes = `07 40` that is 07 * 128 = 896 + h40=64 TOTAL = 960. The Most significant byte is always multiplied by 128. In sysex if you need the number 128 it is 01 00.

I am guessing this is the case for your drum machine, but I haven't read the manual.

[R8m example_1_0.zip](https://github.com/user-attachments/files/16637677/R8m.example_1_0.zip)

**Reply by hps2109** [reply]
Hi John,
many thanks for your reply, your description and your demo panel!!!
I learned so much from you! But unfortunately not everything :-)
Your function looks perfect, so I will check it out with your panel and my "beginner" pitch function.

All mods and yes the pitch mod too, sends correct sysex values with:
Knowing that it has to be "z5" was a challenge :-)
Pitch mod: F0 41 00 36 12 01 24 02 LS MS z5 F7 with slider contents from 0 up to 960: 
where "0" = mod value 480

... and so on
-20=16382
-10=16383
0=0
10=1
20=2
... and so on

The problem is, that the R-8m doesn't sends parameter values. So I must request a full patch dump.
CTRLR receives roundabout 1,2 kbyte in 5 packs with 266 bytes and one pack with 140 bytes.
All values are spliited in nibbles.
So I mangaged it, to combine and set the values for all mods and the pitch value too with my "beginner" function :-)
Also the higher pitch bits "8 and 9 " are swapped over two bytes... 

-- Pitch
   local getBit03 = parameterData1:getByte(112+step)      -- get nibble bit 0-3 0000xxxx (lsb)
   local getBit47 = parameterData1:getByte(113+step)      -- get nibble bit 4-7 0000yyyy
   local getBit8 = parameterData1:getByte(115+step)        -- get nibble incl. bit 8    0000z000
   local getBit9 = parameterData1:getByte(117+step)        -- get nibble incl. bit 9    0000w000 (msb)
   pitchBit8 = bit.rshift(tonumber(getBit8),3)                     -- shift bit 8 to  0000000z
   pitchBit9 = bit.rshift(tonumber(getBit9),3)                     -- shift bit 9 to  0000000w
   local pitchValue = assemble4bitNibble(getBit47, getBit03)     -- Function assemble4bitNibble

   if pitchBit9 == 0 then
      pitchValue = pitchValue + 480 + pitchBit8 * 256
      -- console("Pitch Mod Value+ :"..pitchValue)
      panel:getModulatorByName(selectPitch):setModulatorValue(pitchValue, true, false, true)
  elseif pitchBit9 == 1 then
     pitchValue = pitchValue + 480 + pitchBit8 * 256 - 512
     -- console("Pitch Mod Value- :"..pitchValue)
     panel:getModulatorByName(selectPitch):setModulatorValue(pitchValue, true, false, true)
end

The panel is for me awesome, all mods sends correct sysex and if receiving the bulks, all values are set correctly.
Now I just have to set the missing mods for the keys in this panel.

I have an additional question, if I may ask you?!

KInd regards
Peter

**Comment by dnaldoog** [comment]
Hi Peter,

So have you got it working with negatives?


What is the sysex for -1?

5F 03

or 

7F 7F

???

**Reply by hps2109** [reply]
Yup, it's working :-)
lsb, msb
It's 7F   7F   for -1,   7E,   7f for -2   down to   20   7C   for -480

**Comment by dnaldoog** [comment]
Okay - I know what to do - stand by :)

**Comment by dnaldoog** [comment]
Okay I think this is it:

I didn't really read the question properly.

[incoming midi R-8m_2.zip](https://github.com/user-attachments/files/16643091/incoming.midi.R-8m_2.zip)

This is the algorithm you need to pass all two byte (negative) values through:

<pre>
    local rawValue = (msb * 128) + lsb
    local val = rawValue > 8191 and rawValue - 16384 or rawValue
</pre>

![image](https://github.com/user-attachments/assets/e60ff9d7-5ccc-4736-9431-d3dec15d73be)

**Reply by hps2109** [reply]
Wow, thanks a lot john. The functions are perfect! I'll add it to the panel.

**Reply by hps2109** [reply]
I have a question...
Is it possible with the uiComboBoxes to program a dependency of the respective content of the uiCombo „Media Select“?

The R8m has three PCM-Card Slots, the Cards are „hardware coded“ numbered from 1 to 11 and contain various samples.
The R8m recognizes the PCMCard number and offers the content of this card via media select, so for each key I can set a sample
from „Off“, Internal“ and PCMCard1 perhaps to PCMCard3.

My idea is, to use the information I set in the uiCombos PCM Card1, PCMCard2 & PCMCard3
and set this to the uiCombo Media for each Instrument Key:

Example:
uiCombo PCM Card1 = SN-R8-07 MALLET
uiCombo PCM Card2 = SN-R8-10 DANCE
uiCombo PCM Card3 = SN-R8-03 SOUND EFFECTS

and the uiCombo Media then looks like:
OFF
INTERNAL
SN-R8-07
SN-R8-10
SN-R8-03

or

Example:
uiCombo PCM Card1 = SN-R8-11 METALLIC PERCUSSION
uiCombo PCM Card2 = SN-R8-10 DANCE
uiCombo PCM Card3 = **** EMTY ****

and the uiCombo Media then looks like:
OFF
INTERNAL
SN-R8-11 METALLIC PERCUSSION
SN-R8-10 DANCE

uiCombos PCM Card 1 - 3 with selectable content:
**** EMTY ****
SN-R8-01 CONTEMPORARY PERCUSSION
SN-R8-02 JAZZ BRUSH
SN-R8-03 SOUND EFFECTS
SN-R8-04 ELECTRONIC
SN-R8-05 JAZZ
SN-R8-06 ETHNIC PERCUSSION
SN-R8-07 MALLET
SN-R8-08 DRY
SN-R8-09 POWER DRUMS U.S.A.
SN-R8-10 DANCE
SN-R8-11 METALLIC PERCUSSION

The setting of a modulator content, I’ve understand, but not how to program a dependency of various contents of uiCombos
panel:getComponent(selectMedia):setProperty("uiComboContent","OFF\nINTERNAL\nSN-R8-03“, true)

I hope I expressed it somewhat understandably

Kind regards
Peter

**Comment by dnaldoog** [comment]
Hi Peter,

Do you mean you have one uiCombo with the names of all the cards that the R-8m can accept.

When you select one of those options another uiCombo is repopulated with all the patches from the selected card?

**Reply by hps2109** [reply]
Hi John,
yes, exactly like that!
I'm experiment with a panel. Looks really good so far with "inserted" PCM card 1, but I can't combine the uiCombo values 
"Media" & Instruments" with the other two PCM Cards 2-3. So that these are also available for a selection.

The 13 entries in the uiCombo "Media" are needed, because with this uiCombo, the uiCombo "Instruments" were set, when a SysEx dump is received.
The R-8m doesn't have a parameter, where we can request the actually installed PCM-Cards.
The eventual "PCM-Card Numbers 1-11" are only transmitted within a complete dump for all key# instruments, if they use the card.

Please, have look at the panel.
Kind regards
Peter

**Reply by hps2109** [reply]
CTRL Version: 5.6.30
[Ctrlr Panel (uiCombo, PCMCard -> Media Select -> Key# Instruments)_1_0_2024-08-18_16-24.panel.zip](https://github.com/user-attachments/files/16649049/Ctrlr.Panel.uiCombo.PCMCard.-.Media.Select.-.Key.Instruments._1_0_2024-08-18_16-24.panel.zip)

**Comment by dnaldoog** [comment]
I downloaded you panel - it looks like you have already achieved what you are trying to do?

**Reply by hps2109** [reply]
But unfortunately only for one PCM slot.
At the moment I have no idea, how 3 PCM slots could be linked. Maybe via a large matrix.
However, it's just an idea.

**Comment by dnaldoog** [comment]
You probably need to use [tables](https://www.lua.org/pil/2.5.html) with lists of each set to load. It's central to lua programming anyway.

In my opinion uicombos are unstable, at least in 5.3.201 so I try to use Popups for complex stuff.

Take a look [here](https://sourceforge.net/projects/jv-1010-editor/) at what can be done, but it's a lot more complex to code!

It’s difficult to understand what you want to do. If you can rephrase the question more succinctly I can maybe help.

**Reply by hps2109** [reply]
Thank you very much for the links! I'll try it.

**Comment by dnaldoog** [comment]
Let me know if you have any questions.

**Reply by hps2109** [reply]
Hi John,
first of all, thank you very much for your tip regarding the “table” functions!
Wow, that works really well!

With the table function, I can delete the current position of the three uiButtons and insert the current values in the table.
And... printed out to the console it's perfect and in the correct order, cool! :-)
That's exactly how I wanted it.
If PCM-Card 04 has been selected two or three times, it will only be inserted once! Sure, that makes sense! :-)

But I have two problems,
"Off" and "INTERNAL" must always be retained, but when I select one or up to three card slot with "***EMTY***",
the „INTERNAL" entry is deleted?!?
I've tried several  different functions, if..then..else.. >0, or ==0 and so on, but somehow, I can't find the right solution for it.

example1:
PCMCard1 Value:9
PCMCard2 Value:10
PCMCard3 Value:6

PCMCard1 Name :SN-R8-09
PCMCard2 Name :SN-R8-10
PCMCard3 Name :SN-R8-06

uiButton Media:
OFF
INTERNAL
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
SN-R8-06
**EMTY**
**EMTY**
SN-R8-09
SN-R8-10
**EMTY**

example2:
PCMCard1 Value:9
PCMCard2 Value:10
PCMCard3 Value:11

PCMCard1 Name :SN-R8-09
PCMCard2 Name :SN-R8-10
PCMCard3 Name :SN-R8-11

uiButton Media:
OFF
INTERNAL
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
SN-R8-09
SN-R8-10
SN-R8-11

example3:
PCMCard1 Value:1
PCMCard2 Value:10
PCMCard3 Value:11

PCMCard1 Name :SN-R8-01
PCMCard2 Name :SN-R8-10
PCMCard3 Name :SN-R8-11

uiButton Media:
OFF
INTERNAL
SN-R8-01
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
SN-R8-10
SN-R8-11

wrong: „INTERNAL“ is deleted
PCMCard1 Value:1
PCMCard2 Value:10
PCMCard3 Value:0

PCMCard1 Name :SN-R8-01
PCMCard2 Name :SN-R8-10
PCMCard3 Name :**EMTY**

uiButton Media:
OFF
**EMTY**
SN-R8-01
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
SN-R8-10
**EMTY**

wrong: „INTERNAL“ is deleted
PCMCard1 Value:0
PCMCard2 Value:0
PCMCard3 Value:0

PCMCard1 Name :**EMTY**
PCMCard2 Name :**EMTY**
PCMCard3 Name :**EMTY**

uiButton Media:
OFF
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**
**EMTY**

The second problem is…
How can I insert and correct the table strings to the uiButton format?
At the moment, I have no idea… :-(

—table
mediaSelect = {"OFF", "INTERNAL", "**EMTY**", "**EMTY**“…

—uiButton
panel:getComponent(selectMedia):setProperty("uiComboContent","OFF\nINTERNAL\nSN-R8-01\nSN-R8-02\nSN-R8-03\n…

I would be very happy, if you could give me a tip for my old head :-)
Kind regards
Peter


selectPCMCards2 = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- This variable stops index issues during panel bootup
	if panel:getRestoreState() == true or panel:getProgramState() == true then
		return
	end

--  For instance, if a is the array {10, 20, 30}, after the call table.insert(a, 1, 15) a will be {15, 10, 20, 30}
--  syntax: table.insert(someTable, index, value)
--  The value is inserted at index. If index is not provided, value is inserted at the end of someTable.
--  The function returns the modified, original table.

    -- uiButton mediaSelect with 11 entries

    modIndex = 24 -- Key C1 #24
    local selectMedia = string.format("mediaSelect%1d", modIndex)          -- get mod name mediaSelect + index
    local selectInstrument = string.format("instrument%1d", modIndex)      -- get mod name instrument + index

    local insertCard1 = panel:getModulatorByName("pcmCard1"):getValue()    -- get uiButton current modulator value#
    local insertCard2 = panel:getModulatorByName("pcmCard2"):getValue()
    local insertCard3 = panel:getModulatorByName("pcmCard3"):getValue()

    console("PCMCard1 Value:"..insertCard1)                                -- print uiButton current modulator value#
    console("PCMCard2 Value:"..insertCard2)
    console("PCMCard3 Value:"..insertCard3)

    console("\n")

--  mediaSelect with correct PCM-Card positions
    mediaSelect = {"OFF", "INTERNAL", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**", "**EMTY**"}

    pcmCard1 = {[0]="**EMTY**", "SN-R8-01", "SN-R8-02", "SN-R8-03", "SN-R8-04", "SN-R8-05", "SN-R8-06", "SN-R8-07", "SN-R8-08", "SN-R8-09", "SN-R8-10", "SN-R8-11"} -- array starting at 0
    pcmCard2 = {[0]="**EMTY**", "SN-R8-01", "SN-R8-02", "SN-R8-03", "SN-R8-04", "SN-R8-05", "SN-R8-06", "SN-R8-07", "SN-R8-08", "SN-R8-09", "SN-R8-10", "SN-R8-11"}
    pcmCard3 = {[0]="**EMTY**", "SN-R8-01", "SN-R8-02", "SN-R8-03", "SN-R8-04", "SN-R8-05", "SN-R8-06", "SN-R8-07", "SN-R8-08", "SN-R8-09", "SN-R8-10", "SN-R8-11"}

    console("PCMCard1 Name :"..pcmCard1[insertCard1])                     -- print uiButton content "PCM-Card names"
    console("PCMCard2 Name :"..pcmCard2[insertCard2])
    console("PCMCard3 Name :"..pcmCard3[insertCard3])

    console("\n")

    table.remove(mediaSelect, insertCard1+2)                                            -- remove PCMCard1 at position insertCard1 value
    table.insert(mediaSelect, insertCard1+2, pcmCard1[insertCard1])       -- syntax: table.insert(table, index, value)
    table.remove(mediaSelect, insertCard2+2)                                           -- remove PCMCard2 at position insertCard2 value
    table.insert(mediaSelect, insertCard2+2, pcmCard1[insertCard2])     -- syntax: table.insert(table, index, value)
    table.remove(mediaSelect, insertCard3+2)                                           -- remove PCMCard3 at position insertCard3 value
    table.insert(mediaSelect, insertCard3+2, pcmCard1[insertCard3])      -- syntax: table.insert(table, index, value)

    console("uiButton Media:")
    for i,v in ipairs(mediaSelect) do                                         end

--  example for correct syntax for uiComboContent!
--  panel:getComponent(selectMedia):setProperty("uiComboContent","OFF\nINTERNAL\nSN-R8-01\nSN-R8-02\nSN-R8-03\n**EMTY**\n**EMTY**\n**EMTY**\n**EMTY**\n**EMTY**\n**EMTY**\n**EMTY**\n**EMTY**", true)

--  wrong -> uiComboContent="" !
--  panel:getComponent(selectMedia):setProperty("uiComboContent", mediaSelect, true)

end

**Reply by damiensellier** [reply]
"EMPTY" ?

**Reply by hps2109** [reply]
Hi Damien,
oops... spelling mistake. I corrected it :-)
I was able to correct the problem with the value "INTERNAL" within the table. :-)
But perhaps you can show me, how can I insert the table strings to the uiButton string format?

From
    mediaSelect = {"OFF\n", "INTERNAL\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n", "*EMPTY*\n"}
to
panel:getComponent(selectMedia):setProperty("uiComboContent","OFF\nINTERNAL\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*\n*EMPTY*", true)

panel:getComponent(selectMedia):setProperty("uiComboContent", mediaSelect, true)
doesn't work...
the uiComboContent is then empty: uiComboContent=""

**Comment by dnaldoog** [comment]
Hi Peter,

I am glad you sorted out the previous problem. Smoke was coming out my ears trying to decipher it!

Here is code to populate a `uiButton`

<pre>
TOT=11
fillButton = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local t = {}
    for i = 1, TOT do
        table.insert(t, mediaSelect[i])
        selectMedia:setProperty("uiButtonContent", table.concat(t, "\n"), false)
    end
end

ms = {}
ms.__index = function(t, k)
    t[k] = "(" .. k .. ") EMPTY"
    return t[k]
end

mediaSelect = {"OFF", "INTERNAL"}
setmetatable(mediaSelect, ms)
</pre>

This is a good situation in which to use a rather more advanced lua feature called `metatables`;

but you can also do something like:

<pre>
fillButton = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
selectMedia:setProperty("uiButtonContent",  table.concat({"OFF", "INTERNAL","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"","EMPTY"},"\n"),false)
end
</pre>

 **Simpler is better unless you want to reuse code, in which case, for me,  the metatable option becomes easier to manage in fact**

----
[Fill Button_1_0.zip](https://github.com/user-attachments/files/16833416/Fill.Button_1_0.zip)

**Reply by hps2109** [reply]
Hi John,
thank you very much for all your information.
Of course I didn't want the smoke coming out of your ears...
Ah “table.concat” is the magic function. Many thanks for this info!
Attached is my little demo panel, showing how far I've come at the moment. Perhaps, better than words...

[Ctrlr Panel (PCMCard -> Media Select -> Key# Instruments)_1_0_2024-09-03_17-24.panel.zip](https://github.com/user-attachments/files/16851087/Ctrlr.Panel.PCMCard.-.Media.Select.-.Key.Instruments._1_0_2024-09-03_17-24.panel.zip)

**Reply by dnaldoog** [reply]
The code looks very neat and orderly and your are using tables efficiently I think! Well done!

**Reply by hps2109** [reply]
Thank you very much for your help!!!

**Reply by bolshed** [reply]
Hi, is this a forgotten project or there is something usable for the R-8M?

**Reply by hps2109** [reply]
In principle, the app is ok as far as all the parameters are concerned.
With a few limitations:
e.g., MIDI channel
Unfortunately, I have hard-coded the MIDI channel for "all" parameters (over 1000) to CH1… I had completely forgotten...
The firmware on my R8m seems to be acting up with some of the "Feel" parameters. The values ​​are no longer accepted once a certain value is reached. But I forgot which ones.

I also couldn't get the dependency regarding inserted PCM cards and the mediaSelect button to work.

**Reply by bolshed** [reply]
Thanx for the reply. Will try it. Maybe I try making my own.
